base/MAPL_GridManager.F90

```diff
@@ -78,14 +79,83 @@ subroutine add_prototype(this, grid_type, prototype)
       
    end subroutine add_prototype
 
+   ! Is prototype_name present in the prototypes map keys?
+   logical function is_valid_prototype(this, prototype_name, unusable, rc)
+      class (GridManager), intent(inout) :: this
+      character(len=*), intent(in) :: prototype_name
+      class (KeywordEnforcer), optional, intent(in) :: unusable
+      integer, optional, intent(out) :: rc
+      integer :: status
+      logical, save :: initialized = .false.
+
+      _UNUSED_DUMMY(unusable)
+
+      ! This is a local variable to prevent the function from running
+      ! initialiazation twice. It avoids a shared state variable for
+      ! initialization status.
+      initialized = .false.
+
+      ! Do not initialize prototypes more than once.
+      if (.not. initialized) then
+         call initialize_prototypes(this, _RC)
+         initialized = .true.
+      end if
+
+      is_valid_prototype = .not. this%prototypes%find(prototype_name) == this%prototypes%end()
+
+      _RETURN(_SUCCESS)
+
+   end function is_valid_prototype
+
+   ! Add (name, GridFactory) items to prototypes map.
+   !---------------
+   ! Note:
+   ! We need to add LatLon prototype somewhere, and MAPL does not have
+   ! a natural initialization.  Other grids can be added during
+   ! setServices or initialize of the component that defines the grid.
+   !---------------
+   subroutine initialize_prototypes(this, unusable, rc)
+      use MAPL_LatLonGridFactoryMod, only: LatLonGridFactory
+      use MAPL_CubedSphereGridFactoryMod, only: CubedSphereGridFactory
+      use MAPL_TripolarGridFactoryMod, only: TripolarGridFactory
+      use MAPL_LlcGridFactoryMod, only: LlcGridFactory
+      use MAPL_ExternalGridFactoryMod, only: ExternalGridFactory
+
+      class (GridManager), intent(inout) :: this
+      class (KeywordEnforcer), optional, intent(in) :: unusable
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      type (LatLonGridFactory) :: latlon_factory
+      type (CubedSphereGridFactory) :: cubed_factory
+      type (TripolarGridFactory) :: tripolar_factory
+      type (LlcGridFactory) :: llc_factory
+      type (ExternalGridFactory) :: external_factory
+ 
+      ! This is a local variable to prevent the subroutine from running
```

Comment:
It is late in the day, but ... I know this logic must already exist somewhere.   I am concerned that something has been duplicated.   You probably need to walk me through this PR.    (Not complaining - just want to understand.)

Reply:
I agree. In the current code, it exists in the `make_clone` function. I moved it to a new auxiliary subroutine to be more accessible, but I don't think that's the best place for it. Having said that, I don't know where the best place is.

Reply:
The `make_clone()` procedure explicitly listed the 5 concrete factories I highlighted?   That seems surprising.

Reply:
I was just tired and missed the corresponding deletes.